File: CulinarySiteApp/src/app/services/image.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';

@Injectable()
export class ImageService {
  public url: string = '/api/image';

  constructor(private http: HttpClient) {}

  /*public getImageList(): Observable<IImage[]> {
    return this.http.get<IImage[]>(this.url);
  }

  public getImage(id: number): Observable<IImage> {
    return this.http.get<IImage>(this.url + '/' + id);
  }

  public createImage(image: Image): Observable<void> {
    return this.http.post<void>(this.url, image);
  }

  public updateImage(image: Image): Observable<void> {
    return this.http.put<void>(this.url, image);
  }

  public delete(id: number): Observable<void> {
    return this.http.delete<void>(this.url + '/' + id);
  }*/

  // public getDishImageList(
  //   withRelated: boolean
  // ): Observable<IDishImageListModel[]> {
  //   return this.http.get<IDishImageListModel[]>(
  //     this.url + '/' + withRelated + '/' + 'dish'
  //   );
  // }

  // public getDishImage(
  //   id: number,
  //   withRelated: boolean
  // ): Observable<IDishDetailModel> {
  //   return this.http.get<IDishDetailModel>(
  //     this.url + '/' + id + '/' + withRelated + '/' + 'dish'
  //   );
  // }

  // public createDishImage(
  //   createDishImageModel: CreateDishImageModel
  // ): Observable<void> {
  //   return this.http.post<void>(this.url + '/' + 'dish', createDishImageModel);
  // }

  // public updateDishImage(
  //   updateDishImageModel: UpdateDishImageModel
  // ): Observable<void> {
  //   return this.http.put<void>(this.url + '/' + 'dish', updateDishImageModel);
  // }

  // public getEpisodeImageList(
  //   withRelated: boolean
  // ): Observable<IEpisodeImageListModel[]> {
  //   return this.http.get<IEpisodeImageListModel[]>(
  //     this.url + '/' + withRelated + '/' + 'episode'
  //   );
  // }

  // public getEpisodeImage(
  //   id: number,
  //   withRelated: number
  // ): Observable<IEpisodeDetailModel> {
  //   return this.http.get<IEpisodeDetailModel>(
  //     this.url + '/' + id + '/' + withRelated + '/' + 'episode'
  //   );
  // }

  // public createEpisodeImage(
  //   createEpisodeImageModel: CreateEpisodeImageModel
  // ): Observable<void> {
  //   return this.http.post<void>(
  //     this.url + '/' + 'episode',
  //     createEpisodeImageModel
  //   );
  // }

  // public updateEpisodeImage(
  //   updateEpisodeImageModel: UpdateEpisodeImageModel
  // ): Observable<void> {
  //   return this.http.put<void>(
  //     this.url + '/' + 'episode',
  //     updateEpisodeImageModel
  //   );
  // }

  // public deleteImage(id: number): Observable<void> {
  //   return this.http.delete<void>(this.url + '/' + id);
  // }

  // public getRecipeImageList(
  //   withRelated: boolean
  // ): Observable<IRecipeImageListModel[]> {
  //   return this.http.get<IRecipeImageListModel[]>(
  //     this.url + '/' + withRelated + '/' + 'recipe'
  //   );
  // }

  // public getRecipeImage(
  //   id: number,
  //   withRelated: number
  // ): Observable<IRecipeDetailModel> {
  //   return this.http.get<IRecipeDetailModel>(
  //     this.url + '/' + id + '/' + withRelated + '/' + 'recipe'
  //   );
  // }

  // public createRecipeImage(
  //   createRecipeImageModel: CreateRecipeImageModel
  // ): Observable<void> {
  //   return this.http.post<void>(
  //     this.url + '/' + 'recipe',
  //     createRecipeImageModel
  //   );
  // }

  // public updateRecipeImage(
  //   updateRecipeImageModel: UpdateRecipeImageModel
  // ): Observable<void> {
  //   return this.http.put<void>(
  //     this.url + '/' + 'recipe',
  //     updateRecipeImageModel
  //   );
  // }
}

File: CulinarySite/CulinarySite.Bll/Services/ImageService.cs
using System.Collections.Generic;
using AutoMapper;
using CulinarySite.Bll.Interfaces;
using CulinarySite.Common.Dtos.Image.DishImage;
using CulinarySite.Common.Dtos.Image.EpisodeImage;
using CulinarySite.Common.Dtos.Image.RecipeImage;
using CulinarySite.Dal.Interfaces;
using CulinarySite.Domain.Entities;
using System.Linq;

namespace CulinarySite.Bll.Services
{
    public class ImageService : IImageService
    {
        private readonly IReadOnlyGenericRepository<Image> _imageReadOnlyRepository;
        private readonly IWriteGenericRepository<Image> _imageWriteRepository;
        private readonly IMapper _mapper;
        public ImageService(
            IReadOnlyGenericRepository<Image> imageReadOnlyRepository,
            IWriteGenericRepository<Image> imageWriteRepository,
            IMapper mapper)
        {
            _imageReadOnlyRepository = imageReadOnlyRepository;
            _imageWriteRepository = imageWriteRepository;
            _mapper = mapper;
        }

        public void CreateDishImage(CreateDishImageDto createDishImageDto)
        {
            Image image = _mapper.Map<Image>(createDishImageDto);

            _imageWriteRepository.Create(image);
            _imageWriteRepository.Save();
        }

        public void UpdateDishImage(UpdateDishImageDto updateDishImageDto)
        {
            Image image = _mapper.Map<Image>(updateDishImageDto);

            _imageWriteRepository.Update(image);
            _imageWriteRepository.Save();
        }

        public void DeleteImage(int id)
        {
            _imageWriteRepository.Delete(id);
            _imageWriteRepository.Save();
        }

        public IEnumerable<DishImageListDto> GetDishImageList(bool withRelated)
        {
            IEnumerable<Image> images;
            IEnumerable<DishImageListDto> dishImageListDtos;

            if (withRelated)
            {
                images = _imageReadOnlyRepository.GetItemListWithInclude(x => x.Dish);
                dishImageListDtos = images.Select(x => _mapper.Map<DishImageListDto>(x));
                
                return dishImageListDtos;
            }

            images = _imageReadOnlyRepository.GetItemList();
            dishImageListDtos = images.Select(x => _mapper.Map<DishImageListDto>(x));

            return dishImageListDtos;
        }

        public DishImageDetailDto GetDishImage(int id, bool withRelated)
        {
            var image = new Image();
            var dishImageDetailDto = new DishImageDetailDto();

            if (withRelated)
            {
                image = _imageReadOnlyRepository.GetItemWithInclude(
                    x => x.Id == id,
                    x => x.Dish);

                dishImageDetailDto = _mapper.Map<DishImageDetailDto>(image);

                return dishImageDetailDto;
            }

            image = _imageReadOnlyRepository.GetItem(id);

            dishImageDetailDto = _mapper.Map<DishImageDetailDto>(image);

            return dishImageDetailDto;
        }

        public void CreateEpisodeImage(CreateEpisodeImageDto createEpisodeImageDto)
        {
            Image image = _mapper.Map<Image>(createEpisodeImageDto);

            _imageWriteRepository.Create(image);
            _imageWriteRepository.Save();
        }

        public void UpdateEpisodeImage(UpdateEpisodeImageDto updateEpisodeImageDto)
        {
            Image image = _mapper.Map<Image>(updateEpisodeImageDto);

            _imageWriteRepository.Update(image);
            _imageWriteRepository.Save();
        }

        public IEnumerable<EpisodeImageListDto> GetEpisodeImageList(bool withRelated)
        {
            IEnumerable<Image> images;
            IEnumerable<EpisodeImageListDto> episodeImageListDtos ;

            if (withRelated)
            {
                images = _imageReadOnlyRepository.GetItemListWithInclude(x => x.Episode);
                episodeImageListDtos = images.Select(x => _mapper.Map<EpisodeImageListDto>(x));                

                return episodeImageListDtos;
            }

            images = _imageReadOnlyRepository.GetItemList();
            episodeImageListDtos = images.Select(x => _mapper.Map<EpisodeImageListDto>(x));

            return episodeImageListDtos;
        }

        public EpisodeImageDetailDto GetEpisodeImage(int id, bool withRelated)
        {
            var image = new Image();
            var episodeImageDetailDto = new EpisodeImageDetailDto();

            if (withRelated)
            {
                image = _imageReadOnlyRepository.GetItemWithInclude(
                    x => x.Id == id,
                    x => x.Episode);

                episodeImageDetailDto = _mapper.Map<EpisodeImageDetailDto>(image);

                return episodeImageDetailDto;
            }

            image = _imageReadOnlyRepository.GetItem(id);

            episodeImageDetailDto = _mapper.Map<EpisodeImageDetailDto>(image);

            return episodeImageDetailDto;
        }

        public void CreateRecipeImage(CreateRecipeImageDto createRecipeImageDto)
        {
            Image image = _mapper.Map<Image>(createRecipeImageDto);

            _imageWriteRepository.Create(image);
            _imageWriteRepository.Save();
        }

        public void UpdateRecipeImage(UpdateRecipeImageDto updateRecipeImageDto)
        {
            Image image = _mapper.Map<Image>(updateRecipeImageDto);

            _imageWriteRepository.Update(image);
            _imageWriteRepository.Save();
        }

        public IEnumerable<RecipeImageListDto> GetRecipeImageList(bool withRelated)
        {
            IEnumerable<Image> images;
            IEnumerable<RecipeImageListDto> recipeImageListDtos ;

            if (withRelated)
            {
                images = _imageReadOnlyRepository.GetItemListWithInclude(x => x.Recipe);
                recipeImageListDtos=images.Select(x=> _mapper.Map<RecipeImageListDto>(x));                

                return recipeImageListDtos;
            }

            images = _imageReadOnlyRepository.GetItemList();
            recipeImageListDtos = images.Select(x => _mapper.Map<RecipeImageListDto>(x));

            return recipeImageListDtos;
        }

        public RecipeImageDetailDto GetRecipeImage(int id, bool withRelated)
        {
            var image = new Image();
            var recipeImageDetailDto = new RecipeImageDetailDto();

            if (withRelated)
            {
                image = _imageReadOnlyRepository.GetItemWithInclude(
                    x => x.Id == id,
                    x => x.Recipe);

                recipeImageDetailDto = _mapper.Map<RecipeImageDetailDto>(image);

                return recipeImageDetailDto;
            }

            image = _imageReadOnlyRepository.GetItem(id);

            recipeImageDetailDto = _mapper.Map<RecipeImageDetailDto>(image);

            return recipeImageDetailDto;
        }
    }
}
